### PR TITLE
Support torch.xpu.mem_get_info API

### PR DESCRIPTION
**Pull request by guangyey**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141230

# Motivate
Fix https://github.com/pytorch/pytorch/issues/130599
This PR intends to add a new API, `torch.xpu.mem_get_info,` which is widely used in popular model workloads.
For example, [here](https://github.com/huggingface/accelerate/blob/403c0714d1dd019a481022afce5df75a9963ecd9/src/accelerate/utils/modeling.py#L721) we need to get current GPU memory usage to split or load the model.

cc @gujinghui @EikanWang @fengyuan14

### CODE REVIEW

**Review by dvrogozh — COMMENTED** [review]
We have open issue for this API. Please, add to description that it fixes https://github.com/pytorch/pytorch/issues/130599

**Review by EikanWang — APPROVED** [review]
LGTM

File: torch/csrc/xpu/Module.cpp [hunk 1]
```diff
@@ -274,6 +274,26 @@ PyObject* THXPModule_resetAccumulatedMemoryStats(
   Py_RETURN_NONE;
 }
 
+PyObject* THXPModule_getMemoryInfo(PyObject* self, PyObject* arg) {
+  HANDLE_TH_ERRORS
+#if SYCL_COMPILER_VERSION >= 20250000
```
[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:279]: yes

File: torch/csrc/xpu/Module.cpp [hunk 2]
```diff
@@ -274,6 +274,26 @@ PyObject* THXPModule_resetAccumulatedMemoryStats(
...
+  const auto device_index = THPUtils_unpackDeviceIndex(arg);
+  auto total = at::xpu::getDeviceProperties(device_index)->global_mem_size;
+  auto free = c10::xpu::get_raw_device(device_index)
+                  .get_info<sycl::ext::intel::info::device::free_memory>();
+  auto ret = THPObjectPtr{PyTuple_New(2)};
+  PyTuple_SET_ITEM(ret.get(), 0, THPUtils_packUInt64(free));
+  PyTuple_SET_ITEM(ret.get(), 1, THPUtils_packUInt64(total));
+  return ret.release();
+#else
+  TORCH_CHECK_NOT_IMPLEMENTED(
```
[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:290]: User could use `torch.version.xpu` to query it.

**Review by EikanWang — APPROVED** [review]

File: torch/csrc/xpu/Module.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (dvrogozh) [reply torch/csrc/xpu/Module.cpp:279]: @guangyey : how to install oneapi pytorch dependencies with the compiler version 2025? It's not yet available in the apt channel described at https://www.intel.com/content/www/us/en/developer/articles/tool/pytorch-prerequisites-for-intel-gpu/2-5.html.

[the remark this answers is not in the document]
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp:279]: Currently, you could use `apt-get install -y intel-deep-learning-essentials-2025.0` to install the basekit that used in CI.

**Comment by guangyey** [comment]
We have filed an issue to track this unrelated failure, refer to https://github.com/pytorch/pytorch/issues/141861

**Comment by EikanWang** [comment]
@albanD , @ezyang , @malfet , @atalman , may I know if you have comments regarding this PR?

File: torch/xpu/memory.py [hunk 3]
```diff
@@ -178,10 +178,27 @@ def max_memory_reserved(device: _device_t = None) -> int:
...
+def mem_get_info(device: _device_t = None) -> Tuple[int, int]:
+    r"""Return the global free and total GPU memory for a given device.
+
+    Args:
+        device (torch.device or int or str, optional): selected device. Returns
+            statistic for the current device, given by :func:`~torch.xpu.current_device`,
+            if :attr:`device` is ``None`` (default).
+
+    Returns:
+        int: the memory avialble on the device in units of bytes.
```
Comment (ezyang) [inline torch/xpu/memory.py:190]: available

Comment (guangyey) [reply torch/xpu/memory.py]: @ezyang Updated. Thanks.

File: torch/csrc/xpu/Module.cpp [hunk 4]
```diff
@@ -274,6 +274,26 @@ PyObject* THXPModule_resetAccumulatedMemoryStats(
   Py_RETURN_NONE;
 }
 
+PyObject* THXPModule_getMemoryInfo(PyObject* self, PyObject* arg) {
```
Comment (albanD) [inline torch/csrc/xpu/Module.cpp:277]: nit: use pybind11 for this kind of functions. It makes the code simpler to maintain.

**Review by albanD — APPROVED** [review]
SGTM !

File: torch/csrc/xpu/Module.cpp (hunk 4 above)
Comment (guangyey) [reply torch/csrc/xpu/Module.cpp]: Thanks, @albanD, pybind11 makes my life easier. We will file another PR to bind these methods left in this file via pybind11.

**Pull request closed without merging**